src/topp/GNPSExport.cpp

```diff
@@ -80,296 +150,387 @@ class TOPPGNPSExport : public TOPPBase
+    setValidStrings_("output_type", ListUtils::create<String>("full_spectra,merged_spectra,most_intense"));
 
-    registerTOPPSubsection_("merged_spectra", "Options for exporting mgf file with merged spectra per feature");
+    registerTOPPSubsection_("merged_spectra", "Options for exporting mgf file with merged spectra per consensusElement");
+    registerDoubleOption_("merged_spectra:precursor_mass_tolerance", "<num>", DEF_PREC_MASS_TOL, "Precursor mass tolerance (Da) for ms annotations", false);
     registerDoubleOption_("merged_spectra:cos_similarity", "<num>", DEF_COSINE_SIMILARITY, "Cosine similarity threshold for merged_spectra output", false);
   }
 
   // the main function is called after all parameters are read
-  ExitCodes main_(int, const char **) override
+  ExitCodes main_(int, const char**) override
   {
-    ProgressLogger progress_logger;
-    progress_logger.setLogType(log_type_);
-
     //-------------------------------------------------------------
     // parsing parameters
     //-------------------------------------------------------------
+    // int max_msmap_cache(getIntOption_("msmap_cache"));
+    int pept_cutoff(getIntOption_("peptide_cutoff"));
+
+    double cos_sim_threshold(getDoubleOption_("merged_spectra:cos_similarity"));
+    double bin_width(getDoubleOption_("ms2_bin_size"));
+
     String consensus_file_path(getStringOption_("in_cm"));
     StringList mzml_file_paths(getStringList_("in_mzml"));
-
     String out(getStringOption_("out"));
-
     String output_type(getStringOption_("output_type"));
+    
 
-    double prec_mz_tol(getDoubleOption_("precursor_mz_tolerance"));
-    double prec_rt_tol(getDoubleOption_("precursor_rt_tolerance"));
-
-    double cos_sim(getDoubleOption_("merged_spectra:cos_similarity"));
+    ofstream output_file(out);
 
     //-------------------------------------------------------------
     // reading input
     //-------------------------------------------------------------
     // ConsensusMap
     ConsensusXMLFile consensus_file;
-    consensus_file.setLogType(log_type_);
     ConsensusMap consensus_map;
     consensus_file.load(consensus_file_path, consensus_map);
 
-    // MSExperiment
-    vector<MSExperiment> ms_maps;
-    for (const auto& mzml_file_path : mzml_file_paths)
-    {
-      MzMLFile mzml_file;
-      MSExperiment map;
-      mzml_file.setLogType(log_type_);
-      mzml_file.load(mzml_file_path, map);
 
-      ms_maps.push_back(map);
-    }
+    //-------------------------------------------------------------
+    // preprocessing: allocate memory
+    //-------------------------------------------------------------
+    // max_msmap_cache = std::min(max_msmap_cache, static_cast<int>(mzml_file_paths.size()));
+    int max_msmap_cache = static_cast<int>(mzml_file_paths.size());
+    MzMLFile* mzml_files = new MzMLFile[max_msmap_cache];
+    MSExperiment* specs_list = new MSExperiment[max_msmap_cache];
```

Comment:
please avoid manual heap allocations and use `vector<MSExperiment>`

Reply:
i felt that it would be easier allocate memory for all the MSExperiments in the heap, and the reason that I maintained it as an array rather than wrapping it in a vector was to allow for quicker queries.

Instead, I can opt to wrap it as a vector<MSExperiment>(max_msmap_cache) as I may allow a flag to adjust how many maps to cache as it does consume memory when running the module on larger datasets that require 1000s of mzml files.

Would it be passable to commit this change to opting to wrap the array as a vector?

Reply:
Also because the consensusXML may not necessarily reference all mzml files passed in

Reply:
Vector also allocates its data on the heap. So speedwise, there is no difference whatsoever.
If you preallocate the vector using the same size as the array, it behaves exactly the same as the array.

Reply:
> 
> 
> Also because the consensusXML may not necessarily reference all mzml files passed in

This argues pro vector actually, if you use push_back instead of preallocating and `operator[]`.

But an empty MSExperiment is nothing to worry about memory-wise, so i'd go for the latter.